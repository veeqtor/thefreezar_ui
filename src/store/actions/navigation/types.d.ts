import { ActionType } from 'typesafe-actions';
import * as actions from '.';
import { IApplicationRootState } from '../../../types';

type NavigationActions = ActionType<typeof actions>;

interface INavigationState {
  readonly data: Record<string, unknown>;
  readonly nextPageRoute: string;
  readonly status: {
    readonly isChanged: boolean;
    readonly isRetrieved: boolean;
    readonly isRetrieving: boolean;
  };
}

type RootState = IApplicationRootState;
type ContainerState = INavigationState;
type ContainerActions = NavigationActions;

export { RootState, ContainerState, ContainerActions };
