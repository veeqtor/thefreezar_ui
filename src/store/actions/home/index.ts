import { createAction } from 'typesafe-actions';
import { HomeTypes } from 'store/actions/home/home.constants';
import { Payload } from 'types';

const homePageRetrieve = createAction(HomeTypes.HOME_RETRIEVE_REQUEST)<undefined>();
const homePageRetrieveSuccess = createAction(HomeTypes.HOME_RETRIEVE_SUCCESS)<Payload>();
const homePageRetrieveError = createAction(HomeTypes.HOME_RETRIEVE_FAILED)<Payload>();

export { HomeTypes, homePageRetrieve, homePageRetrieveError, homePageRetrieveSuccess };
