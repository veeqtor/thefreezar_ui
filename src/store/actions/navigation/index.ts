import { createAction } from 'typesafe-actions';
import { NavigationTypes } from './navigation.constants';
import { Payload } from '../../../types';

const retrieveLastPageState = createAction(NavigationTypes.NAVIGATION_RETRIEVE_REQUEST)<undefined>();
const savePageState = createAction(NavigationTypes.NAVIGATION_RETRIEVED)<Payload>();
const goToNextPage = createAction(NavigationTypes.NAVIGATION_CHANGE)<Payload>();

export { NavigationTypes, goToNextPage, retrieveLastPageState, savePageState };
