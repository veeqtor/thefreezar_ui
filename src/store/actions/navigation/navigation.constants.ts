export enum NavigationTypes {
  NAVIGATION_RETRIEVE_REQUEST = '@@naviagtion/RETRIEVE_REQUEST',
  NAVIGATION_RETRIEVED = '@@naviagtion/RETRIEVED',
  NAVIGATION_CHANGE = '@@naviagtion/CHANGE',
}
