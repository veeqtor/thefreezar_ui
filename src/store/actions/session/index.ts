import { createAction } from 'typesafe-actions';
import { SessionTypes } from 'store/actions/session/session.constant';
import { Payload } from 'types';

const sessionPageRetrieve = createAction(SessionTypes.SESSION_RETRIEVE_REQUEST)<undefined>();
const sessionPageRetrieveSuccess = createAction(SessionTypes.SESSION_RETRIEVE_SUCCESS)<Payload>();
const sessionPageRetrieveError = createAction(SessionTypes.SESSION_RETRIEVE_FAILED)<Payload>();

export { SessionTypes, sessionPageRetrieve, sessionPageRetrieveSuccess, sessionPageRetrieveError };
