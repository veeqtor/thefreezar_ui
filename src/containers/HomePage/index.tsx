import * as React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import SEO from '../../components/layouts/SEO';
import { goToNextPage, retrieveLastPageState } from '../../store/actions/navigation';
import { IApplicationRootState } from '../../types';

interface IHomepageProps {
  title: string;
}

const HomePage = ({ title }: IHomepageProps): React.ReactElement => {
  const [click, toggleClick] = React.useState(false);
  const dispatch = useDispatch();
  const currentLocation = useSelector((state: IApplicationRootState) => state.router.action);
  React.useEffect((): void => {
    if (click) {
      dispatch(retrieveLastPageState());
      dispatch(goToNextPage({ nextPageRoute: '/path' }));
    }
  });

  return (
    <div>
      <SEO title={title} />
      <h1>Welcome Home!! {title}</h1>
      <button onClick={(): void => toggleClick(!click)}>Click me</button>
      <p>{currentLocation}</p>
    </div>
  );
};

export default HomePage;
