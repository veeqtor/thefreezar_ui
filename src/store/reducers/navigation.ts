import { NavigationTypes } from '../actions/navigation';
import { ContainerState, ContainerActions } from '../actions/navigation/types';

const initialState: ContainerState = {
  data: {},
  nextPageRoute: '',
  status: {
    isChanged: false,
    isRetrieved: false,
    isRetrieving: false,
  },
};

export default function navigationReducer(
  state: ContainerState = initialState,
  action: ContainerActions,
): ContainerState {
  switch (action.type) {
    case NavigationTypes.NAVIGATION_RETRIEVE_REQUEST:
      return {
        ...state,
        status: {
          ...state.status,
          isRetrieving: true,
        },
      };

    case NavigationTypes.NAVIGATION_RETRIEVED:
      return {
        ...state,
        data: {
          ...state.data,
          ...action.payload.data,
        },
        status: {
          ...state.status,
          isRetrieving: false,
          isRetrieved: true,
        },
      };

    case NavigationTypes.NAVIGATION_CHANGE:
      return {
        ...state,
        data: {
          ...state.data,
          ...action.payload.data,
        },
        nextPageRoute: action.payload.nextPageRoute,
        status: {
          ...state.status,
          isRetrieved: false,
          isChanged: true,
        },
      };

    default:
      return state;
  }
}
