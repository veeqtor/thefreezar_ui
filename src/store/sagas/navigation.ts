/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { takeLatest, all, fork, put, select } from 'redux-saga/effects';
import { push } from 'connected-react-router';

import { NavigationTypes, savePageState } from '../actions/navigation';

export function* nextPage(action: any) {
  yield put(push(action.payload.nextPageRoute));
}

export function* loadCurrentPage() {
  const getRouter = (state: any) => state.router;
  const router = yield select(getRouter);
  yield put(savePageState({ data: router.location.state }));
}

export function* watchLoadCurrentPage() {
  yield takeLatest(NavigationTypes.NAVIGATION_RETRIEVE_REQUEST, loadCurrentPage);
}

export function* watchNextPage() {
  yield takeLatest(NavigationTypes.NAVIGATION_CHANGE, nextPage);
}
export default function* navigationSaga() {
  yield all([fork(watchNextPage), fork(watchLoadCurrentPage)]);
}
