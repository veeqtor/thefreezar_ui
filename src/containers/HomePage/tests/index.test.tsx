/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/no-explicit-any */

import * as React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';

import HomePage from '../index';
import { Provider } from 'react-redux';
import * as navigationActions from '../../../store/actions/navigation';
import configureStore from '../../../store/config';

jest.mock('../../../store/actions/navigation');
const renderHomePage = (store: any, props: any) => {
  return {
    ...render(
      <Provider store={store}>
        <HomePage {...props} />
      </Provider>,
    ),
  };
};

describe('Test Init', () => {
  let store: any;
  const mockedgoToNextPage = navigationActions.goToNextPage as jest.Mock;
  const mockedretrieveLastPageState = navigationActions.retrieveLastPageState as jest.Mock;

  beforeAll(() => {
    mockedgoToNextPage.mockImplementation(() => ({ type: '' }));
    mockedretrieveLastPageState.mockImplementation(() => ({ type: '' }));
  });

  beforeEach(() => {
    store = configureStore('test');
    mockedretrieveLastPageState.mockClear();
    mockedgoToNextPage.mockClear();
  });

  afterEach(cleanup);

  it('Should render welcome text', () => {
    const {
      getByText,
      container: { firstChild },
    } = renderHomePage(store, { title: 'Alejandro' });
    expect(getByText('Welcome Home!! Alejandro')).toBeInTheDocument();
    expect(firstChild).toMatchSnapshot();
  });

  it('Should call the redux dispatches', () => {
    const { getByText } = renderHomePage(store, { title: 'Alejandro' });
    fireEvent.click(getByText('Click me'));
    expect(navigationActions.goToNextPage).toHaveBeenCalled;
    expect(navigationActions.retrieveLastPageState).toHaveBeenCalled;
  });
});
