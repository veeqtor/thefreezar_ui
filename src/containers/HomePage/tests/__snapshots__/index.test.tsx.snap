// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Init Should render welcome text 1`] = `
<div>
  <h1>
    Welcome Home!! 
    Alejandro
  </h1>
  <button>
    Click me
  </button>
  <p>
    POP
  </p>
</div>
`;
